### PR TITLE
Send frame 802.11 over WiFi

### LINKED ISSUE #44: Implement transmission of 802.11 frames on WifiNode
The method should have a signature like `void sendPacket(ipv4_headers_t* headers, void* payload, size_t length);`

### PR DESCRIPTION
**Pull request by IvoAP**
Closes #44

### CODE REVIEW

**Review by pxds — CHANGES_REQUESTED** [review]
Please don't submit commented code for PR :)

File: platformio.ini [hunk 1]
```diff
@@ -15,10 +15,10 @@ build_flags =
 	${env.build_flags}
 	-DMONITOR_CLASS=DisplayMonitor
 	-DBLINK_LED=2
-	-DGATEWAY_SSID='"My GW Network (change me)"'
-	-DGATEWAY_SSID_PASSWORD='"my p4ssw0rd! (change me)"'
-upload_port = COM5
-monitor_port = COM5
+	-DGATEWAY_SSID='"NET_2G45C3B2"'
+	-DGATEWAY_SSID_PASSWORD='"B345C3B2"'
```
Comment (alexfrt) [inline platformio.ini:19]: Is this your home network SSID and password?

File: src/main.cpp [hunk 2]
```diff
@@ -46,14 +46,15 @@ void setup() {
...
+  // mesh = new LoraMesh(wifiToLoraQueue, loraToWifiQueue, new Router());
 
-  if (shouldEnableGateway(GATEWAY_SSID, GATEWAY_SSID_PASSWORD)) {
-    wifi = new Gateway(GATEWAY_SSID, GATEWAY_SSID_PASSWORD, wifiToLoraQueue,
-                       loraToWifiQueue);
-  } else {
-    wifi = new AccessPoint(ACCESS_POINT_SSID, wifiToLoraQueue, loraToWifiQueue);
-  }
+  // if (shouldEnableGateway(GATEWAY_SSID, GATEWAY_SSID_PASSWORD)) {
+  wifi = new Gateway(GATEWAY_SSID, GATEWAY_SSID_PASSWORD, wifiToLoraQueue,
```
Comment (alexfrt) [inline src/main.cpp:52]: There's no need to change these lines to enable gateway. Use the platformio.ini configurations to do that.

File: src/wifinode.cpp [hunk 3]
```diff
@@ -26,7 +28,7 @@ wifi_node_status_t WifiNode::getStatus() {
 }
 
 void WifiNode::sendFragmentationNeeded(ipv4_headers_t* sourcePacket) {
-  struct raw_pcb* pcb = raw_new(IP_PROTO_ICMP);
+  struct raw_pcb* pcb = raw_new(IP_PROTO_TCP);
```
Comment (alexfrt) [inline src/wifinode.cpp:31]: Why did you change this from ICMP to TCP?

File: src/wifinode.cpp [hunk 4]
```diff
@@ -52,3 +54,16 @@ void WifiNode::sendFragmentationNeeded(ipv4_headers_t* sourcePacket) {
   pbuf_free(pbuff);
   raw_remove(pcb);
 }
+
+void WifiNode::sendPacket(ipv4_headers_t* headers,
+                          void* payload,
+                          size_t length) {
+  struct raw_pcb* protocolControlBlock = raw_new(headers->protocol);
+  struct pbuf* pbuff = pbuf_alloc(PBUF_IP, *(uint16_t*)payload, PBUF_RAM);
```
Comment (alexfrt) [inline src/wifinode.cpp:62]: Are you sure that the second argument is a pointer to payload?

**Review by alexfrt — CHANGES_REQUESTED** [review]

File: src/gateway.cpp [hunk 5]
```diff
@@ -83,9 +83,11 @@ void Gateway::routeToInternet() {
   auto message = txQueue->poll();
   if (message == nullptr)
     return;
-
   auto l2Data = &message->data.layer2;
-  printLayer2Data(l2Data);  // TODO route to Internet instead of print...
+  auto ipv4 = (ipv4_headers_t*)l2Data->payload;
+  void* payload = malloc(l2Data->length);
+  memcpy(payload, l2Data->payload, l2Data->length);
```
Comment (alexfrt) [inline src/gateway.cpp:89]: Why is it needed to copy the payload to a new memory buffer?

File: src/main.cpp [hunk 6]
```diff
@@ -7,6 +7,9 @@
 #include "monitor.h"
 #include "wifinode.h"
 
+// Just to the test
+#include "lwip/arch.h"
```
Comment (alexfrt) [inline src/main.cpp:11]: Why this include?

File: src/wifinode.cpp [hunk 7]
```diff
@@ -52,3 +54,18 @@ void WifiNode::sendFragmentationNeeded(ipv4_headers_t* sourcePacket) {
   pbuf_free(pbuff);
   raw_remove(pcb);
 }
+
+void WifiNode::sendPacket(ipv4_headers_t* headers,
+                          void* payload,
+                          size_t length) {
+  struct raw_pcb* protocolControlBlock = raw_new(headers->protocol);
```
Comment (alexfrt) [inline src/wifinode.cpp:61]: What if the `headers->protocol` value is different than ICMP, TCP or UDP?

File: src/wifinode.cpp [hunk 8]
```diff
@@ -52,3 +54,18 @@ void WifiNode::sendFragmentationNeeded(ipv4_headers_t* sourcePacket) {
...
+  uint16_t* data = (uint16_t*)malloc(length);
+  memcpy(data, payload, length);
+  struct pbuf* pbuff = pbuf_alloc(PBUF_IP, *data, PBUF_RAM);
```
Comment (alexfrt) [inline src/wifinode.cpp:64]: Are you sure `pbuf_alloc` is being called with correct arguments?

**Review by alexfrt — CHANGES_REQUESTED** [review]

File: src/gateway.cpp (hunk 5 above)
Comment (PietroAndrade) [reply src/gateway.cpp:89]: Because we think this memory address is going to be emptied. That way, we move on as a copy to be safe.

File: src/main.cpp (hunk 6 above)
Comment (PietroAndrade) [reply src/main.cpp:11]: It will be removed, it was just a test

File: src/wifinode.cpp (hunk 7 above)
Comment (PietroAndrade) [reply src/wifinode.cpp:61]: It is a good question! I am thinking of a solution to this problem

File: src/gateway.cpp (hunk 5 above)
Comment (alexfrt) [reply src/gateway.cpp:89]: It is not going to be freed before transmission. This copy is not necessary.

File: src/main.cpp [hunk 9]
```diff
@@ -46,14 +46,15 @@ void setup() {
   wifiToLoraQueue = new DataQueue<message_t>(DATA_QUEUE_LENGTH);
   loraToWifiQueue = new DataQueue<message_t>(DATA_QUEUE_LENGTH);
 
-  mesh = new LoraMesh(wifiToLoraQueue, loraToWifiQueue, new Router());
+  // mesh = new LoraMesh(wifiToLoraQueue, loraToWifiQueue, new Router());
```
[the remark this answers is not in the document]
Comment (IvoAP) [reply src/main.cpp]: We already removed

File: src/main.cpp [hunk 10]
```diff
@@ -46,14 +46,15 @@ void setup() {
...
-  } else {
-    wifi = new AccessPoint(ACCESS_POINT_SSID, wifiToLoraQueue, loraToWifiQueue);
-  }
+  // if (shouldEnableGateway(GATEWAY_SSID, GATEWAY_SSID_PASSWORD)) {
+  wifi = new Gateway(GATEWAY_SSID, GATEWAY_SSID_PASSWORD, wifiToLoraQueue,
+                     loraToWifiQueue);
+  // } else {
+  //   wifi = new AccessPoint(ACCESS_POINT_SSID, wifiToLoraQueue,
+  //   loraToWifiQueue);
+  // }
```
[the remark this answers is not in the document]
Comment (IvoAP) [reply src/main.cpp]: We already removed

File: src/main.cpp (hunk 2 above)
Comment (IvoAP) [reply src/main.cpp]: We already removed this comments, and we are using the platformio ini. Thanks!

File: platformio.ini [hunk 11]
```diff
@@ -15,10 +15,10 @@ build_flags =
...
+; upload_port = COM5
+; monitor_port = COM5
```
[the remark this answers is not in the document]
Comment (IvoAP) [reply platformio.ini]: We already removed.

File: platformio.ini (hunk 1 above)
Comment (IvoAP) [reply platformio.ini]: We already corret this.
https://github.com/Instituto-Atlantico/mesh-connect/blob/ivo/transmission80211wifinode/platformio.ini

File: src/main.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (IvoAP) [reply src/main.cpp]: https://github.com/Instituto-Atlantico/mesh-connect/blob/ivo/transmission80211wifinode/src/main.cpp

File: src/wifinode.cpp (hunk 3 above)
Comment (IvoAP) [reply src/wifinode.cpp]: We already fixed.

Comment (IvoAP) [reply src/wifinode.cpp]: https://github.com/Instituto-Atlantico/mesh-connect/blob/ivo/transmission80211wifinode/src/wifinode.cpp

File: src/main.cpp (hunk 6 above)
Comment (IvoAP) [reply src/main.cpp]: It's already removed

File: src/wifinode.cpp (hunk 7 above)
Comment (IvoAP) [reply src/wifinode.cpp:65]: It's already checked:
 if (headers->protocol != IP_PROTO_ICMP && headers->protocol != IP_PROTO_UDP &&
      headers->protocol != IP_PROTO_TCP) {
    return;
  }

File: src/wifinode.cpp (hunk 8 above)
Comment (IvoAP) [reply src/wifinode.cpp]: Now we are passing the correct arguments int routeToInternet of gateway.cpp and passing the correct arguments of in pbuff_alloc.

**Review by alexfrt — APPROVED** [review]

**Review by pxds — APPROVED** [review]
